USE ROLE SYSADMIN;
USE DATABASE SAGE;
USE SCHEMA PORTAL_RAW;

CREATE OR REPLACE VIEW SAGE.PORTAL_RAW.PORTAL_MERGE AS (
    SELECT
        "id",
        'ad' AS PORTAL,
        "name",
        "study",
        "dataType",
        "assay",
        "organ",
        "tissue",
        "species",
        "sex",
        "consortium",
        "grant",
        "modelSystemName",
        "treatmentType",
        TO_VARIANT(ARRAY_CONSTRUCT("specimenID")) AS "specimenID",
        "individualID",
        "individualIdSource",
        "specimenIdSource",
        "resourceType",
        "dataSubtype",
        "metadataType",
        "assayTarget",
        "analysisType",
        "cellType",
        "nucleicAcidSource",
        "fileFormat",
        "group",
        "isModelSystem",
        "isConsortiumAnalysis",
        "isMultiSpecimen",
        TO_TIMESTAMP_NTZ("createdOn" / 1000) AS "createdOn",
        "createdBy",
        "parentId",
        "currentVersion",
        "benefactorId",
        "projectId",
        "modifiedOn",
        "modifiedBy",
        "dataFileHandleId",
        "metaboliteType",
        "chromosome",
        "modelSystemType",
        "libraryPrep",
        "dataFileSizeBytes",
        -- ELITE
        NULL AS "type",
        NULL AS "consent",
        NULL AS "libraryType",
        NULL AS "platform",
        NULL AS "project",
        NULL AS "etag",
        -- NF
        NULL AS "isMultiIndividual",
        NULL AS "diagnosis",
        NULL AS "tumorType",
        NULL AS "fundingAgency",
        NULL AS "nf1Genotype",
        NULL AS "nf2Genotype",
        NULL AS "isCellLine",
        NULL AS "studyId",
        NULL AS "age",
        NULL AS "readPair",
        NULL AS "accessType",
        NULL AS "accessTeam",
        NULL AS "modelOf",
        NULL AS "compoundName",
        NULL AS "experimentalCondition",
        NULL AS "isXenograft",
        NULL AS "transplantationType",
        NULL AS "isPairedEnd",
        NULL AS "isStranded",
        NULL AS "libraryPreparationMethod",
        NULL AS "initiative",
        NULL AS "progressReportNumber",
        NULL AS "Resource_id",
        NULL AS "genePerturbationType",
        NULL AS "genePerturbationMethod",
        NULL AS "genePerturbed",
        NULL AS "contentType"
    FROM
        AD
    UNION ALL
    SELECT
        "id",
        'elite' AS PORTAL,
        "name",
        TO_VARIANT(ARRAY_CONSTRUCT("study")) AS "study",
        TO_VARIANT(ARRAY_CONSTRUCT("dataType")) AS "dataType",
        TO_VARIANT(ARRAY_CONSTRUCT("assay")) AS "assay",
        NULL AS "organ",
        NULL AS "tissue",
        TO_VARIANT(ARRAY_CONSTRUCT("species")) AS "species",
        NULL AS "sex",
        "consortium",
        TO_VARIANT(ARRAY_CONSTRUCT("grant")) AS "grant",
        NULL AS "modelSystemName",
        NULL AS "treatmentType",
        NULL AS "specimenID",
        NULL AS "individualID",
        NULL AS "individualIdSource",
        NULL AS "specimenIdSource",
        "resourceType",
        CAST("dataSubtype" AS VARCHAR) AS "dataSubtype",
        "metadataType",
        NULL AS "assayTarget",
        "analysisType",
        NULL AS "cellType",
        NULL AS "nucleicAcidSource",
        "fileFormat",
        NULL AS "group",
        "isModelSystem",
        "isConsortiumAnalysis",
        "isMultiSpecimen",
        NULL AS "createdOn",
        NULL AS "createdBy",
        NULL AS "parentId",
        "currentVersion",
        NULL AS "benefactorId",
        NULL AS "projectId",
        NULL AS "modifiedOn",
        NULL AS "modifiedBy",
        NULL AS "dataFileHandleId",
        NULL AS "metaboliteType",
        NULL AS "chromosome",
        CAST("modelSystemType" AS VARCHAR) AS "modelSystemType",
        CAST("libraryPrep" AS VARCHAR) AS "libraryPrep",
        NULL AS "dataFileSizeBytes",
        "type",
        "consent",
        "libraryType",
        CAST("platform" AS VARCHAR) AS "platform",
        "project",
        "etag",
        -- NF
        NULL AS "isMultiIndividual",
        NULL AS "diagnosis",
        NULL AS "tumorType",
        NULL AS "fundingAgency",
        NULL AS "nf1Genotype",
        NULL AS "nf2Genotype",
        NULL AS "isCellLine",
        NULL AS "studyId",
        NULL AS "age",
        NULL AS "readPair",
        NULL AS "accessType",
        NULL AS "accessTeam",
        NULL AS "modelOf",
        NULL AS "compoundName",
        NULL AS "experimentalCondition",
        NULL AS "isXenograft",
        NULL AS "transplantationType",
        NULL AS "isPairedEnd",
        NULL AS "isStranded",
        NULL AS "libraryPreparationMethod",
        NULL AS "initiative",
        NULL AS "progressReportNumber",
        NULL AS "Resource_id",
        NULL AS "genePerturbationType",
        NULL AS "genePerturbationMethod",
        NULL AS "genePerturbed",
        NULL AS "contentType"
    FROM
        ELITE
    UNION ALL
    SELECT
        "id",
        'NF' AS PORTAL,
        "name",
        TO_VARIANT(ARRAY_CONSTRUCT("studyName")) AS "study",
        TO_VARIANT(ARRAY_CONSTRUCT("dataType")) AS "dataType",
        TO_VARIANT(ARRAY_CONSTRUCT("assay")) AS "assay",
        NULL AS "organ",
        TO_VARIANT(ARRAY_CONSTRUCT("tissue")) AS "tissue",
        TO_VARIANT(ARRAY_CONSTRUCT("species")) AS "species",
        TO_VARIANT(ARRAY_CONSTRUCT("sex")) AS "sex",
        "consortium",
        NULL AS "grant",
        TO_VARIANT(ARRAY_CONSTRUCT("modelSystemName")) AS "modelSystemName",
        NULL AS "treatmentType",
        "specimenID",
        "individualID",
        NULL AS "individualIdSource",
        NULL AS "specimenIdSource",
        "resourceType",
        "dataSubtype",
        NULL AS "metadataType",
        NULL AS "assayTarget",
        NULL AS "analysisType",
        TO_VARIANT(ARRAY_CONSTRUCT("cellType")) AS "cellType",
        NULL AS "nucleicAcidSource",
        "fileFormat",
        NULL AS "group",
        NULL AS "isModelSystem",
        NULL AS "isConsortiumAnalysis",
        TO_BOOLEAN("isMultiSpecimen") AS "isMultiSpecimen", --convert to Bool VARCHAR(16777216),
        TO_TIMESTAMP_NTZ("createdOn" / 1000) AS "createdOn",
        "createdBy",
        "parentId",
        NULL AS "currentVersion",
        "benefactorId",
        "projectId",
        "modifiedOn", -- these should be timestamp
        "modifiedBy",
        NULL AS "dataFileHandleId",
        NULL AS "metaboliteType",
        NULL AS "chromosome",
        NULL AS "modelSystemType",
        "libraryPrep",
        "dataFileSizeBytes",
        "type",
        NULL AS "consent",
        NULL AS "libraryType",
        "platform", -- cast this to string elsewhere VARCHAR(16777216),
        NULL AS "project",
        "etag",
        -- NF only
        TO_BOOLEAN("isMultiIndividual") AS "isMultiIndividual", -- converting VARCHAR(16777216),
        "diagnosis", --Add new VARIANT
        "tumorType", -- add new  VARIANT,
        "fundingAgency", -- add new varchar,
        "nf1Genotype", -- add new VARCHAR(16777216),
        "nf2Genotype", -- add new VARCHAR(16777216),
        "isCellLine", -- add new VARCHAR(16777216),
        "studyId", -- add new VARCHAR(16777216),
        "age", -- add new VARCHAR(16777216), (why is this a string)
        "readPair", -- add new FLOAT,
        "accessType", -- add new VARCHAR(16777216),
        "accessTeam", -- add new FLOAT,
        "modelOf", -- add new VARCHAR(16777216),
        "compoundName", --add new VARCHAR(16777216),
        "experimentalCondition", --add new VARCHAR(16777216),
        "isXenograft", -- add new VARCHAR(16777216), why str
        "transplantationType", -- add new VARCHAR(16777216),
        "isPairedEnd", -- add new VARCHAR(16777216), why str?
        "isStranded", -- add new VARCHAR(16777216), why str?
        "libraryPreparationMethod", -- add new VARCHAR(16777216), is this the same as libraryType
        "initiative", -- new column VARCHAR(16777216),
        "progressReportNumber", -- add new FLOAT,
        "Resource_id", --add new variant,
        "genePerturbationType", -- ad new VARCHAR(16777216),
        "genePerturbationMethod", -- add new NUMBER(38,0),
        "genePerturbed", --add new VARCHAR(16777216),
        "contentType" -- add new VARCHAR(16777216),
    FROM
        NF
);

SELECT * FROM PORTAL_MERGE WHERE PORTAL = 'NF';
